### ISSUE TITLE
Poor billboard quality

### ISSUE THREAD

**Issue by hpinkos** [AnalyticalGraphicsInc/cesium#4235]
Reported on the forum: https://groups.google.com/d/msg/cesium-dev/OLbOuzNRM7U/MHgo_4N6CwAJ

The icon to the left is a billboard, the icon to the right is the original image

![image](https://cloud.githubusercontent.com/assets/3451886/17976183/5d50383a-6aba-11e6-8203-d3608cb913e5.png)

**Comment by emackey** [comment]
I believe this happens because billboards are not aligned to the nearest pixel.  When a billboard has a non-integer screen location, it becomes blurred by the fullscreen AA process.

There's a [suggested fix on Stack Overflow](http://stackoverflow.com/questions/33784256/cesium-label-blurred/33786151#33786151) that involves making a one-line change to [BillboardCollectionVS.glsl#L245](https://github.com/AnalyticalGraphicsInc/cesium/blob/1.24/Source/Shaders/BillboardCollectionVS.glsl#L245), to change it from:

`gl_Position = czm_viewportOrthographic * vec4(positionWC.xy, -positionWC.z, 1.0);`

to:

`gl_Position = czm_viewportOrthographic * vec4(floor(positionWC.xy + 0.5), -positionWC.z, 1.0);`

This has the effect of greatly reducing blur on both Billboards and Labels.

What do you guys think of this?  If you like it, I can make a pull request out of it (assuming we can take code from SO.  I presume most projects already have quite a bit of uncredited SO code around).

The reasoning seems clear: Billboards are always raster images, so non-integer screen locations will always be worse quality than pixel-aligned locations.

**Comment by lilleyse** [comment]
Ah cool, can you post a screenshot of the difference?

**Comment by emackey** [comment]
The Stack Overflow answer has a good screenshot:

http://stackoverflow.com/questions/33784256/cesium-label-blurred/33786151#33786151

**Comment by shunter** [comment]
If this feels like deja-vu, you're right.  Here's where clamp-to-pixel functionality was added: #188  Here's where it was removed again: #565.

**Comment by emackey** [comment]
Good memory @shunter.  In experimenting with this code, it looks like the clamp-to-pixel is not what's actually improving this guy's screenshots.  He's also turning off `scene.fxaa` at the same time, and that appears to be what actually cleans up the label.

**Comment by mramato** [comment]
Yeah, fxaa is probably the real culprit here.  Clamping to pixel actually does not do what we would hope it would do (Especially with text).  It also causes all kinds of problem when during motion (either because the item is moving or the camera is moving).  Not clamping really is the "correct" option.

I asked Dan about possibly not running fxaa on labels and billboards and he things it might be possible but might require too much overhead or other craziness.

I just noticed he's also provided us with an svg, that's probably compounding the problem because the svg is properly be anti-aliased on rasterization by the browser (and depending on the browser it could be doing a bad job).

@hpinkos can you reach out and get the original icon?  I didn't notice it in that thread.

**Comment by emackey** [comment]
I asked @bagnell offline how our fxaa is actually implemented.  He suggested this was based on an older approximation where high-contrast edges are intentionally blurred, which is fast enough for realtime, but is a very different implementation from true anti-aliasing.  There was talk of this possibly being upgraded to a different algorithm when WebGL 2 becomes more widely available.

So it sounds like the blurring is a design feature of the current fxaa system, making it almost unavoidable without a rewrite or disabling of fxaa.

**Comment by pjcozzi** [comment]
I don't think it is out of the question to render to one framebuffer that would get FXAA, and another for labels/billboards/points that would not, and then composite them.  This would also work well with the need to render labels on top of polygons since the low level-renderer now would have explicit-ish knowledge of labels/billboards/points.

This will be a non-trivial amount of time though so we can't do it for a few months.

**Comment by hpinkos** [comment]
Here is a png and svg version of the icon from the user.  He says the png looks better than the svg

![icon-sourdough-library 2x](https://cloud.githubusercontent.com/assets/3451886/18054877/b102db4c-6dd3-11e6-883d-009682d032b0.png)
[icon-library.svg.txt](https://github.com/AnalyticalGraphicsInc/agi-cesium-cloud/files/442601/icon-library.svg.txt)

**Comment by mramato** [comment]
> Here is a png and svg version of the icon from the user. He says the png looks better than the svg

This is exactly what I assumed a large part of the problem was.  Browsers do not all do an equally good job of rasterizing/resizing SVGs, so the fact that it's an SVG is one of the reasons it looked so blurry.

**Comment by emackey** [comment]
True, but we still have big problem with fxaa intentionally blurring our billboards and labels.  I think @pjcozzi's comment is the path to persue, when time permits.

**Comment by mramato** [comment]
I think everyone agrees @emackey I just wanted to point out that SVGs can compound the issue.

**Comment by denverpierce** [comment]
Related to #2752 and #3279 .  If you're considering separate frame buffers, being able to selectively apply it to imagery tiles as well would be helpful.  An AA pass on tiles, especially those with labels, reduces quality.

**Comment by hpinkos** [comment]
Also reported here: https://groups.google.com/forum/?hl=en#!topic/cesium-dev/NBfMvj80vGI

**Comment by bagnell** [comment]
I investigated this at the bug bash. It would be quite a bit of work to selectively apply anti-aliasing. Here is the post process that would happen **after each frustum** instead of on the entire scene:
![20170209_122413](https://cloud.githubusercontent.com/assets/1494815/22798399/e29b7650-eecf-11e6-92c4-122fb2ce1c86.jpg)

**Comment by pjcozzi** [comment]
@bagnell can you post a link to the branch-in-progress and before/after screenshots?

**Comment by bagnell** [comment]
The branch is [selective-fxaa](https://github.com/AnalyticalGraphicsInc/cesium/tree/selective-fxaa). 
Before:
![image](https://cloud.githubusercontent.com/assets/1494815/22799612/74f2bcb2-eed4-11e6-9e33-cc0287083ede.png)
After:
![image](https://cloud.githubusercontent.com/assets/1494815/22799624/7be275b2-eed4-11e6-9eea-9349ddc6165c.png)

**Comment by pjcozzi** [comment]
Another idea is to render pick ids with MRT (or a separate pass to an offscreen framebuffer/texture), and then only blur pixels where the surrounding pick ids are different so only the ends of objects are blurred.  Billboards, the globe texture, etc. would not be blurred except at the silhouettes.

**Comment by denverpierce** [comment]
Would these fixes apply to #3279 as well?

**Comment by pjcozzi** [comment]
@denverpierce they would be part of the solution.

**Comment by expatiating** [comment]
I believe that this is still an issue - has there been any additional thought on this topic in the last ~5 years? 

Or is there a recommended approach to solve with billboards and text overlays?

**Comment by ggetz** [comment]
Quality issue when moving reported in https://community.cesium.com/t/billboard-with-custom-canvas-image-gets-darker-brighter-on-scrolling-rotating/25221/2.

**Comment by Huyston** [comment]
I'm still having this issue.
Here is an example: [sandcastle](https://sandcastle.cesium.com/#c=bVRtb9owEP4rFl8SNJpAu00tDWgqbFKlVq2Ktn3JFyc+glXHZrbDS6v+950dQiCqBbLvuefO58fnbKgmGw5b0GRCJGzJDAyvyuiPx8Ig9+ZMSUu5BB30b1OZyg1GmRwkYFAdHXkTnX6OGGRVsVip7S9NSzDPoBeQK8mQb3XV8pY7ShFbUmFakErLqeDUnHjcjrIqM9BPy2fFpXXO0dCNg7fQnC34myvp6vuQxOSR2lVk/mkbngfWR6iDMi5EpqhmLl29+1rzklu+ARNRxsITTe4a8kwJAbnlSob9VpCc4uTyMJVXJUgb5RqohZ8CnIVaeoKX0PF5WZySC7AH5t3+noWBQfU8tw6LtpzZlT90i62AFyvbgL4IvCnY2cs5ggcSJp7VaBhcsmBA3nMllF6saQ5jEhhdZAiu+Q7EL6VLahGsJK6ugw+/f5MyyqDg8hllDc9xY7V6hYXdC6d+2tPA0h4y4rjlCGyfv80RzqI1ShkOBwR/I///LLnfERWLjM7bo1k1p5b+fnlo3a6IiJe0gBeQDDSXha/Jnw9vo64slUulSegUE0oW3FbMlX4xuh7eniAJ6QBfJsc+65P3VBIcbSZqj4luXFhjJ+TMPEmCbXp5TERIM5OTxvRd+N563PDnu2cEb+pIJGvX3cHgEybyUJuOx+RUwN1+zo2lMnecCvVa4jWxLhMLvZePYEGbsXu+Hb/vzPFX9xY7nro/P3WtleHuBeG+h+c1o9riisqraKlVOYdCA5jwKP7gqGD/JJdv0caoVx+p/OgNeonvhWnj/cHLtdKWVFqEURRbKNeuH0ycVfkr4GM1pkmVxKehB4zxDeFskvaEogy76mkDWtB92psmq9H0oQajKEpiNJMY6dNupFVKZFS7kIM/cR8B73PPPe3VWk6+DQ/SuZWvBRlvFxwvaDe+ucEE3dSd7/Rxi/8)
I'm trying to create many billboards with a rectangle texture. I tried supplying an image with the "pixelated" style to prevent the image from blurring. Note that the image on the top is not blurred, but the rendered rectangle in Cesium is blurred.
Also tried disabling fxaa and antialias, without success.

**Comment by dukeofcool199** [comment]
Any further investigation into this issue? I can see that even in the [sandcastle examples](https://sandcastle.cesium.com/?src=Billboards.html) the billboard quality is low as well.

**Comment by ggetz** [comment]
Thanks for the interest. There hasn't been any activity on this item recently, but it is definitely on our radar.

**Comment by dukeofcool199** [comment]
for anyone else that has found this thread, my co-worker came across [this other thread](https://github.com/CesiumGS/cesium/issues/5859) explaining how billboards were made intentionally blurry to help with running on mobile devices. and that viewing billboards at a browser zoom level of anything other than 100% will cause this issue to become more noticeable.
The workaround from the thread states to adjust the value of `viewer.resolutionScale` property
I tested this fix in [this sandcastle](https://sandcastle.cesium.com/?src=Billboards.html) with some success.

**Comment by badg0003** [comment]
@dukeofcool199 Thanks big time for this. I've been searching and made too many attempts to fix this than I want to admit lol, but your tip did the trick! :)

**Comment by CassandraCat** [comment]
> for anyone else that has found this thread, my co-worker came across [this other thread](https://github.com/CesiumGS/cesium/issues/5859) explaining how billboards were made intentionally blurry to help with running on mobile devices. and that viewing billboards at a browser zoom level of anything other than 100% will cause this issue to become more noticeable. The workaround from the thread states to adjust the value of `viewer.resolutionScale` property I tested this fix in [this sandcastle](https://sandcastle.cesium.com/?src=Billboards.html) with some success.
> 
> UPDATE:
> 
> change this setting for the viewer `viewer.useBrowserRecommendedResolution` to be false

nice bro.

**Comment by donmccurdy** [comment]
Here is a small PR affecting the resolution of rasterized SVG content in billboards...

- https://github.com/CesiumGS/cesium/pull/13020

... which changes the result noticeably for the SVG sample [attached above](https://github.com/CesiumGS/cesium/issues/4235#issuecomment-243141721). Scaling is not the root of _this_ issue I think, that may be pixel alignment or anti-aliasing as discussed above, but is a related possible improvement in billboard quality.
